### PR TITLE
4. DZ

### PR DESCRIPTION
**Pull request by Stakljar**

### CODE REVIEW

File: app/src/main/java/agency/five/codebase/android/movieapp/data/database/FavoriteMovieDao.kt [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+
+@Dao
+interface FavoriteMovieDao {
+    @Query("SELECT * FROM DbFavoriteMovie")
+    fun favorites(): Flow<List<DbFavoriteMovie>>
+
+    @Query("DELETE FROM DbFavoriteMovie WHERE id = :movieId")
+    fun deleteMovie(movieId: Int)
+
+    @Insert
```
Comment (abramovic-laura) [inline app/src/main/java/agency/five/codebase/android/movieapp/data/database/FavoriteMovieDao.kt]: Add `OnConflictStrategy.REPLACE` here

File: app/src/main/java/agency/five/codebase/android/movieapp/data/network/model/ApiMovieDetails.kt [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+        releaseDate = releaseDate ?: "",
+        language = originalLanguage,
+        runtime = runtime,
+        crew = crew.map {
+            Crewman(
+                id = it.id,
+                name = it.name,
+                job = it.job
+            )
+        },
```
Comment (abramovic-laura) [inline app/src/main/java/agency/five/codebase/android/movieapp/data/network/model/ApiMovieDetails.kt]: For better readability, you can add `toCrewman` function to `ApiCrew`, as you have for `ApiMovieDetails` and `ApiMovie`, and just call it here. The same can be applied to `ApiCast`

File: app/src/main/java/agency/five/codebase/android/movieapp/data/repository/MovieRepositoryImpl.kt [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+            )
+        }
+    }
+
+    private suspend fun findMovie(movieId: Int): Movie {
+        return movieDetails(movieId).first().movie
+    }
+
+    override suspend fun removeMovieFromFavorites(movieId: Int) {
+        GlobalScope.launch(Dispatchers.IO) {
```
Comment (abramovic-laura) [inline app/src/main/java/agency/five/codebase/android/movieapp/data/repository/MovieRepositoryImpl.kt]: Remove this. Also never use `GlobalScope`.

File: app/src/main/java/agency/five/codebase/android/movieapp/data/repository/MovieRepositoryImpl.kt [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+                    cast = apiMovieCredits.cast
+                )
+            }
+    }.flowOn(bgDispatcher)
+
+    override fun favoriteMovies(): Flow<List<Movie>> = favorites
+
+    override suspend fun addMovieToFavorites(movieId: Int) {
+        val movie = findMovie(movieId)
+        GlobalScope.launch(Dispatchers.IO) {
```
Comment (abramovic-laura) [inline app/src/main/java/agency/five/codebase/android/movieapp/data/repository/MovieRepositoryImpl.kt]: Remove this

File: app/src/main/java/agency/five/codebase/android/movieapp/ui/home/HomeScreen.kt [hunk 5]
```diff
@@ -94,19 +94,79 @@ fun HomeScreen(
...
+                if(_popularCategoryViewState.movies.any { it.id == movieId })
+                    _popularCategoryViewState =
+                        changeMovieFavoriteStatus(_popularCategoryViewState, movieId)
+                if(_nowPlayingCategoryViewState.movies.any { it.id == movieId })
+                    _nowPlayingCategoryViewState =
+                        changeMovieFavoriteStatus(_nowPlayingCategoryViewState, movieId)
+                if(_upcomingCategoryViewState.movies.any { it.id == movieId })
+                    _upcomingCategoryViewState =
+                        changeMovieFavoriteStatus(_upcomingCategoryViewState, movieId)
+            }
```
Comment (abramovic-laura) [inline app/src/main/java/agency/five/codebase/android/movieapp/ui/home/HomeScreen.kt]: There is no need for this in preview, just leave an empty lambda

File: app/src/main/java/agency/five/codebase/android/movieapp/data/repository/MovieRepositoryImpl.kt [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+                DbFavoriteMovie(
+                    id = movie.id,
+                    posterUrl = movie.imageUrl
+                )
+            )
+        }
+    }
+
+    private suspend fun findMovie(movieId: Int): Movie {
+        return movieDetails(movieId).first().movie
```
Comment (abramovic-laura) [inline app/src/main/java/agency/five/codebase/android/movieapp/data/repository/MovieRepositoryImpl.kt]: Search for move in `moviesByCategory`

**Review by abramovic-laura — CHANGES_REQUESTED** [review]

File: app/src/main/java/agency/five/codebase/android/movieapp/data/repository/MovieRepositoryImpl.kt [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+                posterUrl = movie.imageUrl
+            )
+        )
+    }
+
+    private suspend fun findMovie(movieId: Int): Movie {
+        return moviesByCategory.values.asFlow().flattenMerge()
+            .filter { it -> it.firstOrNull { it.id == movieId } != null }
+            .first().first { it.id == movieId }
+    }
```
Comment (abramovic-laura) [inline app/src/main/java/agency/five/codebase/android/movieapp/data/repository/MovieRepositoryImpl.kt:117]: You can just use `moviesByCategory.values.forEach` instead of this `.asFlow().flattenMerge()` and then `.first()`

File: app/src/main/java/agency/five/codebase/android/movieapp/ui/moviedetails/MovieDetailsViewModel.kt [hunk 8]
```diff
@@ -1,24 +1,41 @@
...
+                    overview = "",
+                    isFavorite = false,
+                    crew = listOf(),
+                    cast = listOf()
+                )
+            )
 
     fun toggleFavorite(movieId: Int) {
-        viewModelScope.launch {
+        viewModelScope.launch(Dispatchers.IO) {
```
Comment (abramovic-laura) [inline app/src/main/java/agency/five/codebase/android/movieapp/ui/moviedetails/MovieDetailsViewModel.kt:38]: Why did you add `Dispatchers.IO` to all viewModelScope launches?

**Review by abramovic-laura — CHANGES_REQUESTED** [review]

File: app/src/main/java/agency/five/codebase/android/movieapp/ui/moviedetails/MovieDetailsViewModel.kt (hunk 8 above)
Comment (Stakljar) [reply app/src/main/java/agency/five/codebase/android/movieapp/ui/moviedetails/MovieDetailsViewModel.kt:38]: Can't access db on main thread, getting IllegalStateException.

File: app/src/main/java/agency/five/codebase/android/movieapp/data/repository/MovieRepositoryImpl.kt (hunk 7 above)
Comment (Stakljar) [reply app/src/main/java/agency/five/codebase/android/movieapp/data/repository/MovieRepositoryImpl.kt:117]: Don't know what exactly you mean, how am I supposed to retrieve Movie using forEach?

Comment (abramovic-laura) [reply app/src/main/java/agency/five/codebase/android/movieapp/data/repository/MovieRepositoryImpl.kt:117]: It would look something like this:
```
moviesByCategory.values.forEach {
            val movie = it.first().firstOrNull { movie -> movie.id == movieId }
            if (movie != null) return movie
}
```

File: app/src/main/java/agency/five/codebase/android/movieapp/ui/moviedetails/MovieDetailsViewModel.kt (hunk 8 above)
Comment (abramovic-laura) [reply app/src/main/java/agency/five/codebase/android/movieapp/ui/moviedetails/MovieDetailsViewModel.kt:38]: This shouldn't be an issue, but it is okay to leave it like this

**Review by abramovic-laura — APPROVED** [review]
